src/codewhisperer/tracker/codewhispererCodeCoverageTracker.ts

```diff
@@ -8,88 +8,115 @@ import * as telemetry from '../../shared/telemetry/telemetry'
-        this._acceptedTokens = []
-        this._totalTokens = []
+        this._acceptedTokens = {}
+        this._totalTokens = {}
         this._startTime = 0
         this._language = language
     }
 
-    public setAcceptedTokens(recommendation: string) {
-        const terms = this._globals.get<boolean>(CodeWhispererConstants.termsAcceptedKey) || false
-        if (!terms) return
-
-        // generate accepted recoomendation token and stored in collection
-        this._acceptedTokens.push(...recommendation)
-        this._totalTokens.push(...recommendation)
+    public get acceptedTokens(): { [key: string]: CodeWhispererToken[] } {
+        return this._acceptedTokens
     }
-
-    public get AcceptedTokensLength(): number {
-        return this._acceptedTokens.length
+    public get totalTokens(): { [key: string]: number } {
+        return this._totalTokens
     }
 
-    public setTotalTokens(content: string) {
-        if (this._totalTokens.length === 0 && this._timer == undefined) {
-            const currentDate = new globals.clock.Date()
-            this._startTime = currentDate.getTime()
-            this.startTimer()
-        }
-
-        if (content.length <= 2) {
-            this._totalTokens.push(content)
-        } else if (content.length > 2) {
-            this._totalTokens.push(...content)
-        }
+    public countAcceptedTokens(range: vscode.Range, text: string, filename: string) {
+        const terms = this._globals.get<boolean>(CodeWhispererConstants.termsAcceptedKey) || false
+        if (!terms) return
+        // generate accepted recommendation token and stored in collection
+        this.addAcceptedTokens(filename, { range: range, text: text, accepted: text.length })
+        this.addTotalTokens(filename, text.length)
     }
 
     public flush() {
         const terms = this._globals.get<boolean>(CodeWhispererConstants.termsAcceptedKey) || false
         if (!terms) {
-            this._totalTokens = []
-            this._acceptedTokens = []
+            this._totalTokens = {}
+            this._acceptedTokens = {}
             this.closeTimer()
             return
         }
-        this.emitCodeWhispererCodeContribution()
+        try {
+            this.emitCodeWhispererCodeContribution()
+        } catch (error) {
+            getLogger().error(`Encountered ${error} when emitting code contribution metric`)
+        }
+    }
+
+    public updateAcceptedTokensCount(editor: vscode.TextEditor) {
+        const filename = editor.document.fileName
+        if (filename in this._acceptedTokens) {
+            for (let i = 0; i < this._acceptedTokens[filename].length; i++) {
+                const oldText = this._acceptedTokens[filename][i].text
+                const newText = editor.document.getText(this._acceptedTokens[filename][i].range)
+                this._acceptedTokens[filename][i].accepted =
+                    Math.max(oldText.length, newText.length) - distance(oldText, newText)
```

Comment:
Lei i would suggest we can extract this block of logic to be another function and and we can test against easily. And we can add this test case  https://github.com/aws/aws-toolkit-jetbrains/blob/main/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/CodeWhispererCodeCoverageTrackerTest.kt#L345 for readers to more easily understand what's we wanna achieve here

Reply:
here is ref to the `getAcceptredTokenDelta` for JB https://github.com/aws/aws-toolkit-jetbrains/blob/main/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt#L106

Reply:
More test cases have been added